### PR TITLE
Implement GET /applicants by externalId

### PR DESCRIPTION
**Pull request by bickelj**
It is convenient to check whether applicants already exist before trying to post them (and getting a failure, for example). We use the path parameter `{id}` for PDC internal IDs, not external IDs, so use a query parameter `externalId`

Issue #121 GET /applicants by externalId not working

### CODE REVIEW

**Review by jasonaowen — COMMENTED** [review]
Thanks for working on this, @bickelj!

My main concern is the API design question below.

File: src/database/queries/applicants/selectByExternalId.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+SELECT a.id AS "id",
+  a.external_id as "externalId",
+  a.opted_in AS "optedIn",
+  a.created_at AS "createdAt"
+FROM applicants a
+WHERE a.external_id = :externalId;
```
[the remark this answers is not in the document]
Comment (bickelj) [reply src/database/queries/applicants/selectByExternalId.sql:6]: It's a matter of preference, I suppose. I assume most queries have a join somewhere and then for consistency we use the aliases everywhere. Perhaps this style change could go into a separate PR that changes all of our queries to have a consistent look/feel?

File: src/handlers/applicantsHandlers.ts [hunk 2]
```diff
@@ -54,6 +54,50 @@ const getApplicants = (
     });
 };
 
+const getApplicantsByExternalId = (
+  req: Request,
+  res: Response,
+  next: NextFunction,
+): void => {
+  db.sql('applicants.selectByExternalId', { externalId: req.query.externalId })
```
[the remark this answers is not in the document]
Comment (bickelj) [reply src/handlers/applicantsHandlers.ts:62]: It might be straightforward or might not. If I have a moment, I will try it out.

File: src/__tests__/applicants.int.test.ts [hunk 3]
```diff
@@ -120,6 +162,31 @@ describe('/applicants', () => {
...
+
+    it('returns 503 DatabaseError if an insufficient resources database error is thrown when selecting by externalId', async () => {
+      jest.spyOn(db, 'sql')
+        .mockImplementationOnce(async () => {
+          throw new TinyPgError(
+            'Something went wrong',
+            undefined,
+            {
+              error: {
+                code: PostgresErrorCode.INSUFFICIENT_RESOURCES,
```
[the remark this answers is not in the document]
Comment (bickelj) [reply src/__tests__/applicants.int.test.ts:174]: Not in particular but this error is one example of many potential errors to be thrown from postgres, so it is not intended to cover only this exact case but exercises the code we'd expect to be thrown on the kitchen sink of postgres errors, if I recall correctly.

File: src/__tests__/applicants.int.test.ts [hunk 4]
```diff
@@ -66,6 +66,33 @@ describe('/applicants', () => {
...
+          200,
+          [
+            {
+              createdAt: '2023-04-05T17:40:00.000Z',
+              externalId: '167336',
+              id: 2,
+              optedIn: true,
+            },
+          ],
+        );
```
[the remark this answers is not in the document]
Comment (bickelj) [reply src/__tests__/applicants.int.test.ts:93]: Here is the reasoning:
1. We want a consistent API.
2. We consistently use the `GET /[entity]/{id}` path parameter as PDC-assigned/internal/DB-generated ID. This is guaranteed to return one result in all entities.
3. External ID on applicants is not the same as a PDC-assigned/internal/DB-generated ID. Therefore it is a query parameter.
4. We (will) use query parameters for filtering/searching across all entities.
5. When filtering/searching entities by a query parameter, in general, the result will be many. This particular result happens to be one instead of many.
6. For consistency, we want the response shape to be the same for `GET /applicants` whether we filter or do not filter. We don't want the shape of the response to change with query parameters, just the amount of data or in some narrow cases an optional field present or not present.
7. When a filter/search comes up lacking, one typically would get an empty list instead.

So I agree, but I also already considered that in light of the above points.

File: src/__tests__/applicants.int.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bickelj) [reply src/__tests__/applicants.int.test.ts:174]: But agree with the sentiment.

File: src/__tests__/applicants.int.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasonaowen) [reply src/__tests__/applicants.int.test.ts:93]: I agree on 1-4! However, I don't think 5 or 7 applies, and disagree on 6. Returning a single object or `404`ing IS being consistent - just with `/applicants/:id` rather than with `/applicants?_search={term}`. This, I suppose, is the benefit of separating out `/entities` from `/entity/{id}`, because under that naming convention, this API would be `/applicant?externalId={id}`, making it clear that one and only one result is possible.

Query parameters can be a part of stable URLs, and I don't think we should treat all query parameters the same. A zero- or single-element list is a more cumbersome API to consume than one that gives an object or a 404. Let's do the thing that makes our API easier to consume!

File: src/database/queries/applicants/selectByExternalId.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (jasonaowen) [reply src/database/queries/applicants/selectByExternalId.sql:6]: I opened #321 to start that discussion!

File: src/__tests__/applicants.int.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (slifty) [reply src/__tests__/applicants.int.test.ts:93]: For what it's worth I also would probably interpret attribute-level query params as a filter / search and would expect the response to be the same as other filters in the event of zero matches.  Which is to say, I would be surprised to get a 404 because we provided a search facet that resulted in zero results, and I would expect an empty array.

If we changed to `/applicant?externalId=...` would signal a specific resource attempt, but I also think that breaks my expectations around REST resource names (I expect the resource category to be the same `/applicants` for both single and collection access).  I'm happy to be told my expectations are out of date, but I think they're rooted in at least some historic form of best practice...

If we really want the 404 behavior, what about What about something like `/applicants/externalId/{foo}` which is described [here](https://stackoverflow.com/questions/56969324/rest-api-resource-naming-with-multiple-ids).

The downside of this approach is we now have two distinct URLs for the same resource, whereas with the "filter" approach the collection returned is not pointing to a resource but rather is returning a collection of resources and there is still only one resource URL.

[the remark this answers is not in the document]
Comment (slifty) [reply src/__tests__/applicants.int.test.ts:93]: (oh, I do think this discussion signals that a test case would be reasonable though, to make it clear the intended implementation)

**Comment by bickelj** [comment]
[2aa7ee5](https://github.com/PhilanthropyDataCommons/service/pull/315/commits/2aa7ee554b4898d80eb42a414515a7fe1f107911) rebased on main before making further changes because it has been a minute.

File: src/database/queries/applicants/selectByExternalId.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (bickelj) [reply src/database/queries/applicants/selectByExternalId.sql]: I pushed a change to remove aliasing where it was unneeded, also SHOUTED "AS" MORE CONSISTENTLY! I am OK with shouting SQL but I think it's a funny convention we happened upon.

[the remark this answers is not in the document]
Comment (slifty) [reply src/database/queries/applicants/selectByExternalId.sql]: I never stopped to think about the shouting but you're right 🤣

File: src/handlers/applicantsHandlers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bickelj) [reply src/handlers/applicantsHandlers.ts:62]: I still haven't taken the time to get to this, maybe on an additional pass next week. Extracting a helper sounds easy but for some reason with middleware I think it's more tricky. Regardless, it is probably worth doing to avoid the duplication.

File: src/__tests__/applicants.int.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bickelj) [reply src/__tests__/applicants.int.test.ts:93]: In short, @jasonaowen, I would be surprised or given pause by an API that sometimes returns a single result for query parameters and other times returns a list for query parameters, so I do not think of such an API as easier to consume. I added the test suggested by @slifty.

File: src/__tests__/applicants.int.test.ts [hunk 5]
```diff
@@ -66,6 +66,53 @@ describe('/applicants', () => {
...
+      await db.query(`
+        INSERT INTO applicants (
+          external_id,
+          opted_in,
+          created_at
+        )
+        VALUES
+          ( '3581', 'false', '2023-08-25 08:38:00+0000' ),
+          ( '3583', 'true', '2023-08-25 08:39:00+0000' );
+      `);
```
Comment (jasonaowen) [inline src/__tests__/applicants.int.test.ts:106]: Can we use the `applicants.insertOne` script here instead of directly executing SQL?

File: src/__tests__/applicants.int.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasonaowen) [reply src/__tests__/applicants.int.test.ts:93]: Okay, it seems the main difference is that you are both framing this as a search rather than a lookup. Under that framing, returning an array makes sense!

Reserving the right to later make querying only by `externalId` return more than one response is also a convincing argument, although it's not obvious to me what that'd mean for API versioning/semver - would that be a breaking change? Not something we have to figure out yet, either way.

File: src/handlers/applicantsHandlers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasonaowen) [reply src/handlers/applicantsHandlers.ts:62]: Please do!

**Review by jasonaowen — COMMENTED** [review]
Thanks for moving this forward, @bickelj!

File: src/handlers/applicantsHandlers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bickelj) [reply src/handlers/applicantsHandlers.ts]: Easy enough, it turns out.

File: src/__tests__/applicants.int.test.ts (hunk 5 above)
Comment (bickelj) [reply src/__tests__/applicants.int.test.ts:106]: Technically, yes. But I prefer the test setup code to be directly readable and minimize use of production code in tests other than the code under test.

File: src/__tests__/applicants.int.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bickelj) [reply src/__tests__/applicants.int.test.ts:93]: @jasonaowen Not to beat a dead horse but to acknowledge that this entity is different: in the original API design, I think this was an endpoint that used the natural key (`externalId`) as the URL parameter rather than the synthetic key, but that was an exception among other entities. So we made this one consistent with the synthetic key as a URL parameter. And then yes, I think both @slifty and I think we should consistently treat query parameters as search parameters.

File: src/database/queries/applicants/selectByExternalId.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (reefdog) [reply src/database/queries/applicants/selectByExternalId.sql]: tbh I love the shouting. It's one of those things that makes SQL feel like SQL to me. Maybe because it's "SQL" and not "sql"? 🤔 ← @slifty can we get a :thonking: here plz

**Comment by bickelj** [comment]
165ab67 rebased on main.

File: src/__tests__/applicants.int.test.ts (hunk 5 above)
Comment (jasonaowen) [reply src/__tests__/applicants.int.test.ts:106]: Brevity helps readability! And I would encourage you to re-evaluate that preference: here, we are tightly coupling this test about a search endpoint with knowledge of the underlying structure of the database table. That coupling is inappropriate, as it means that changing the database table will involve changing this unrelated test. Code should have only one reason to change; in this case, the test code should only change if the search endpoint changes.

Comment (bickelj) [reply src/__tests__/applicants.int.test.ts:106]: Adding required columns to the table will require this test to change regardless. Adding optional columns will not require a change to this test regardless. So I think that's unavoidable, right?

Comment (slifty) [reply src/__tests__/applicants.int.test.ts:106]: I do think that it would make a lot of sense to leverage the power of `tinypg` for more concise tests (same way we don't have write SQL in our production code).

I personally would probably default to just having one set of (tested?) `.sql` templates rather than having two copies of `insertApplicant` where one is in `src/db` and the other is in like `src/test/db`, but even those two places would be more DRY than writing raw SQL inline every time.

Thoughts on that approach?

Comment (bickelj) [reply src/__tests__/applicants.int.test.ts:106]: While this review comment seemed like a small change to an existing pull request, it really is opening a new issue regarding our test code practices. I am open to changing our test code practices based on shared goals of test code. I do not think we need to hash it out here.

File: src/database/queries/applicants/selectByExternalId.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (bickelj) [reply src/database/queries/applicants/selectByExternalId.sql]: When y'all are using `psql` do you really hold shift or frequently tap caps lock when querying the database?

[the remark this answers is not in the document]
Comment (reefdog) [reply src/database/queries/applicants/selectByExternalId.sql]: @bickelj I do indeed, truly! (I don't say that as a form of advocacy, just describing my personal pathology.)

File: src/__tests__/applicants.int.test.ts (hunk 5 above)
Comment (jasonaowen) [reply src/__tests__/applicants.int.test.ts:106]: Mixing SQL code as strings in TypeScript code is bad practice. It requires duplication of code, with the all attending issues of making changes to that duplicated code. It makes it harder to read, as it's just a string, rather than clearly marked as SQL, so editors/IDEs are less able to parse it as SQL. It makes it harder to lint, as it's just a string. We're using tinypg to be able to avoid all those problems!

> Adding required columns to the table will require this test to change regardless. Adding optional columns will not require a change to this test regardless. So I think that's unavoidable, right?

That specific example would require a matching change everywhere, yes. However, other changes might not; for example, renaming a table or a required column - but not the associated script name / parameters - would silently succeed when calling the script but require changes to SQL referring to those table/column names directly.

More importantly, making either such change - and reviewing it - will be easier when there's only one calling pattern to update, rather than two. It should be sufficient to search for the name of the SQL script and update each call site, rather than also need to replicate changes in multiple duplicate copies of in-line SQL code.

As we move forward on linting SQL (#321), linting will only ever work on SQL files - mixing SQL code as strings in TypeScript code makes that SQL opaque to our tools.

> While the original review comment seemed like a small change to an existing pull request, it really is opening a new issue regarding our test code practices. I am open to changing our test code practices based on shared goals of test code. I do not think we need to hash it out here.
> 
> It appears the issue might have already been created here: https://github.com/PhilanthropyDataCommons/service/issues/331

That issue is actually for improving our tests beyond the change I've requested here; the link in #331 goes to [a comment on #320](https://github.com/PhilanthropyDataCommons/service/pull/320#discussion_r1183117521) where we're using `applicants.insertOne`. And, in fact, we did hash this out previously: https://github.com/PhilanthropyDataCommons/service/pull/265#discussion_r1144197166 . It was not a long discussion, and we obviously didn't convert all existing SQL-strings-in-test-code as part of that, but I'm also not asking you to convert everything here - just stop introducing new SQL strings.

Comment (slifty) [reply src/__tests__/applicants.int.test.ts:106]: I just wanna +1 that while it would be good to eventually make this fix everywhere, the first step towards improving coding practice / eliminating the tech debt is to make sure *new* contributions adhere to the best practice!

(We already do it both ways, and I think @jasonaowen's points explain why this would be an improvement to inline SQL.)

Comment (bickelj) [reply src/__tests__/applicants.int.test.ts]: [2fa610e](https://github.com/PhilanthropyDataCommons/service/pull/315/commits/2fa610e17fd8a1b6cace8885d115e31fcded7840) rebased on main.

[6d02917](https://github.com/PhilanthropyDataCommons/service/pull/315/commits/6d029176773404a8e92d062c3f27e2d3bc3a9f6e) has the requested change.

TLDR: do not misinterpret the code change here as forever binding assent to a new direction for the other existing tests or future tests. If you have time, read on.

The key readability of tests to me is at the moment of test failure. Put another way, the guiding question is "when a test fails, is it clear why that test failed?" This question is not exactly present for production code so there can be slight differences in the way we write tests versus production code. DRY is important in all code and is the strongest argument I have heard so far for changing the test to be more like the tests changed in https://github.com/PhilanthropyDataCommons/service/pull/265.

I want tests to fail for the right reasons and not the wrong reasons. To that end, a reason I avoid a call to production code (the insertOne code) in the test is to avoid the test failing for some bug in the code not under test here (the insertOne code). If the insert failed using a call to `db.sql` instead, that would be more obviously a bug in the test.

I want tests to have strong assertions. I want those assertions to be simple. I don't want tests to have what Meszaros calls a "Mystery Guest". To those ends, I include all the fields in the row and assert all fields in the row, literally and straightforwardly.

While it is OK to re-use the same data across tests, provided those data are carefully designed, I also find it helpful to have distinct data per test such that if I needed to search for a value in a given (i.e. failing) test's data or search the tests for a given data value (i.e. the db state after some test failures) I would be able to see exactly where that value came from.

So balancing DRY against these other objectives, I lean toward these other objectives more heavily than toward DRY. I acknowledge that DRY is better than not DRY in most circumstances. I am still open to revisiting the excess  test code duplication issue separately from this pull request.

<details><summary>As an example of what I'd like to avoid, when changing the code to be more closely aligned with the changes in #265 (for the sake of getting this code merged), I got this test failure:</summary>

```TypeScript
    // Here is the new code
    it('returns the expected applicant by externalId', async () => {
      await db.sql('applicants.insertOne', {
        externalId: '54280485',
        optedIn: false,
      });
      await db.sql('applicants.insertOne', {
        externalId: '167336',
        optedIn: true,
      });
      await agent
        .get('/applicants?externalId=167336')
        .set(authHeader)
        .expect(
          200,
          [
            {
              createdAt: expectTimestamp,
              externalId: '167336',
              id: 2,
              optedIn: true,
            },
          ],
        );
    });
```

</details>

``` TypeScript
// Here is the test failure
 FAIL  src/__tests__/applicants.int.test.ts (7.216 s)
  ● /applicants › GET / › returns the expected applicant by externalId

    expected [
      {
        createdAt: StringMatching {
          '$$typeof': Symbol(jest.asymmetricMatcher),
          sample: /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/,
          inverse: false
        },
        externalId: '167336',
        id: 2,
        optedIn: true
      }
    ] response body, got [
      {
        id: 2,
        externalId: '167336',
        optedIn: true,
        createdAt: '2023-09-12T16:17:42.084Z'
      }
    ]
```

<details><summary>Now I need to figure out why this test fails. It isn't clear from the above.</summary>Immediately I am given pause by the expected `createdAt` value. It is obscure and obviously different from `'2023-09-12T16:17:42.084Z'`. If I set that aside, I look at the remaining fields and find that they match. The `externalId` matches, the `optedIn` matches, and the `id` matches. This test passed before. Hmm. Perhaps the order of keys in a map is now significant. I go back to the original code and move the `createdAt`. What if I rule out the order by re-arranging it to `id, externalId, optedIn, createdAt`? Still fails. OK it's not order. No, order was not significant before. Is the import of the timestamp matcher the same? Yes. Perhaps the regex pattern does not exactly match. No, the regex looks good. Ah, the `expect` is now broken in two in the code pushed in #265, as [noted by @slifty](https://github.com/PhilanthropyDataCommons/service/pull/265#discussion_r1149553517). If I change the `expect` statements to this, it passes:

```TypeScript
        .expect(200)
        .expect((res) => expect(res.body).toEqual(
// ...
```
</details>

The reason I took the trouble to push back on these comments is I would like an acknowledgment that the change is not without its own cost. I might be misinterpreting, but the previous two comments come across as saying "the current code is bad practice and the suggested change is good practice" rather than recognizing that there are reasons for each approach and tradeoffs involved. One programmer might weigh the balance in one way and another might weigh it another way. We might all agree to goals of concise, readable, consistent, DRY, correctly passing, and correctly failing tests, but the priority order of the goals and our judgments about the long-term consequences of the exact implementation toward those goals might be different.

It was also brought up that "we did hash this out previously", but the "we" there was @slifty and @jasonaowen. My approval of a pull request is a judgment that, on balance, the code is an improvement or neutral and not moving backwards. If @slifty and @jasonaowen agree to something related to that code, it does not forever bind the remaining team members to that agreement. Nor is full agreement of everything in that pull request required for approval of a pull request. It is a different situation when I am asked to change something and after reflection still come to the conclusion that I prefer the original. It is also a different situation when the request is obviously not just about a few lines of code in a pull request but about a change in direction for code outside the request.

I think the benefit of review is improvements to the code. Where those suggested improvements are obvious to the requester, the person who requested review makes them with little trouble. Where those improvements are not obvious, the person doing the implementation should be given deference unless it is a big deal. I don't see this as a crucial, big deal to change these few lines to be on balance something like a neutral or slightly negative change. So I made the requested change here for the sake of getting this code merged.

**Comment by bickelj** [comment]
a8b78e rebased on main again.

**Comment by bickelj** [comment]
With #777 we are taking a different tack such that the `applicants` endpoint is no longer present. See 789ef97bb1c6838046a9abf602097915645a367c and #798. There is no need to throw even more good stones after bad given that this endpoint is gone.
